d3a_interface/data_classes.py

```diff
@@ -0,0 +1,442 @@
+from typing import List, Dict, Optional, Tuple, Union
+from pendulum import DateTime, parse
+
+from d3a_interface.utils import datetime_to_string_incl_seconds, key_in_dict_and_not_none
+
+
+def my_converter(date_obj: DateTime):
+    """Define how to convert datetime objects while serializing to json."""
+    if isinstance(date_obj, DateTime):
+        return date_obj.isoformat()
+    return None
+
+
+class BaseBidOffer:
+    """Base class defining shared functionality of Bid and Offer market structures."""
+    def __init__(self, id: str, time: datetime, price: float, energy: float,
+                 original_price: Optional[float] = None,
+                 attributes: Dict = None, requirements: List[Dict] = None):
+        self.id = str(id)
+        self.time = time
+        self.original_price = original_price or price
+        self.price = price
+        self.energy = energy
+        self.energy_rate = self.price / self.energy
+        self.attributes = attributes
+        self.requirements = requirements
+
+    def update_price(self, price: float) -> None:
+        self.price = price
+        self.energy_rate = self.price / self.energy
+
+    def update_energy(self, energy: float) -> None:
+        self.energy = energy
+        self.energy_rate = self.price / self.energy
+
+    def to_json_string(self, **kwargs) -> str:
+        """Convert the Offer or Bid object into its JSON representation.
+
+        Args:
+            **kwargs: additional key-value pairs to be added to the JSON representation.
+        """
+        obj_dict = deepcopy(self.__dict__)
+        if kwargs:
+            obj_dict = {**obj_dict, **kwargs}
+
+        obj_dict["type"] = self.__class__.__name__
+
+        return json.dumps(obj_dict, default=my_converter)
+
+    def serializable_dict(self) -> Dict:
+        return {
+            "type": self.__class__.__name__,
+            "id": self.id,
+            "energy": self.energy,
+            "energy_rate": self.energy_rate,
+            "original_price": self.original_price,
+            "time": datetime_to_string_incl_seconds(self.time),
+            "attributes": self.attributes,
+            "requirements": self.requirements
+        }
+
+    @classmethod
+    def from_json(cls, offer_or_bid, current_time=None) -> Union["Offer", "Bid"]:
+        offer_bid_dict = json.loads(offer_or_bid)
+        object_type = offer_bid_dict.pop("type")
+
+        offer_bid_dict.pop("energy_rate", None)
+
+        if object_type == "Offer":
+            offer_bid_dict["time"] = current_time
+            return Offer(**offer_bid_dict)
+        if object_type == "Bid":
+            return Bid(**offer_bid_dict)
+
+
+class Offer(BaseBidOffer):
+    def __init__(self, id: str, time: datetime, price: float,
+                 energy: float, seller: str, original_price: Optional[float] = None,
+                 seller_origin: str = None, seller_origin_id: str = None,
+                 seller_id: str = None,
```

Comment:
These parameters should not have a default None and must always be provided

Reply:
I'm not sure, there are dozens of calls to Bid/ Offer instantiation (especially in tests) and they were always omitted
I have checked and it looks like in the production code, these are always sent
Basically I agree that it makes sense to not allow these to be None, however this will require changes to all unit tests, do you think these can be postponed to another time?

Reply:
Okay, please make a ticket to be in radar that this is something to be dealt with.

Reply:
I'll keep the same ticket and move it to SFD after this is merged, and I'll update its description to include more refactorings in the future (I already have some in mind)

Reply:
> I'll keep the same ticket and move it to SFD after this is merged, I'll update its description to include more refactorings in the future (I already have some in mind)

If the description of ticket needs to be updated at the time of review, it means thing weren't planned thought of before starting and its something additional, so I would vote to move this ticket to done once this PR is merged, and create a follow-up ticket targeting further enhancements.

Tagging @fievelk @spyrostz in case they don't agree with me .

Reply:
I'm okay with creating a new ticket, but it will also be a broad : refactor market structures
THe original tickets did not explicitly describe the needed changes indeed, it's a `refactoring` ticket so I resort to my instincts on the go while refactoring, but the majority of changes were mentioned in an older ticket (mostly comments from Pierpaolo) + I added some more changes and left some for the future

Reply:
Since you want to first merge this PR, and then create a new one to address this specific comment, I would say that we can safely open a new ticket. It will also help to track the changes with the commit messages' prefixes.